fw/ja5_conf.h

```diff
@@ -20,19 +20,242 @@
  * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  */
 
+#pragma once
+
+#include <linux/list.h>
+#include <linux/hashtable.h>
+#include <linux/slab.h>
+
 #include "lib/ja5.h"
 #include "cfg.h"
+#include "log.h"
+#include "hash.h"
+#include "tempesta_fw.h"
+
+#ifndef JA5_STRUCT_T
+#error "JA5_STRUCT_T macro is not defined before include of ja5_conf.h"
+#endif
+
+/* Define default size as multiple of TDB extent size */
+#define JA5_DEFAULT_STORAGE_SIZE ((1 << 21) * 25)
+#define JA5_HASHTABLE_BITS 10
+
+typedef struct {
+	struct hlist_node	hlist;
+	atomic_t		refcnt;
+	JA5_STRUCT_T		ja5_hash;
+	u64			conns_per_sec;
+	u64			records_per_sec;
+} Ja5HashEntry;
+
+typedef struct {
+	u64 storage_size;
+	DECLARE_HASHTABLE(hashes, JA5_HASHTABLE_BITS);
+} Ja5FilterCfg;
+
+static Ja5FilterCfg __rcu	*filter_cfg;
+static Ja5FilterCfg		*filter_cfg_reconfig;
+
+static Ja5HashEntry*
+get_ja5_hash_entry(JA5_STRUCT_T fingerprint)
+{
+	u64 key;
+	Ja5HashEntry *entry = NULL;
+	Ja5FilterCfg *cfg;
+
+	if (!filter_cfg)
+		return NULL;
+
+	key = hash_calc((char *)&fingerprint, sizeof(fingerprint));
```

Comment:
Hash from hash that introduce collision, seems radix more applicable here, however it's just note, don't change anything.

Maybe we can use ja5 as key for hashtable without crc hashing?

Reply:
Yeah, maybe it makes sense just to leave a TODO comment here